server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -528,34 +499,64 @@ public String toString() {
     }
 
     /** Notify the master node that the given shard has been successfully snapshotted **/
-    void notifySuccessfulSnapshotShard(final Snapshot snapshot,
-                                       final ShardId shardId,
-                                       final String localNodeId,
-                                       final DiscoveryNode masterNode) {
-        sendSnapshotShardUpdate(snapshot, shardId, new ShardSnapshotStatus(localNodeId, State.SUCCESS), masterNode);
+    private void notifySuccessfulSnapshotShard(final Snapshot snapshot, final ShardId shardId, DiscoveryNode masterNode) {
+        sendSnapshotShardUpdate(
+            snapshot, shardId, new ShardSnapshotStatus(clusterService.localNode().getId(), State.SUCCESS), masterNode);
     }
 
     /** Notify the master node that the given shard failed to be snapshotted **/
-    void notifyFailedSnapshotShard(final Snapshot snapshot,
-                                   final ShardId shardId,
-                                   final String localNodeId,
-                                   final String failure,
-                                   final DiscoveryNode masterNode) {
-        sendSnapshotShardUpdate(snapshot, shardId, new ShardSnapshotStatus(localNodeId, State.FAILED, failure), masterNode);
+    private void notifyFailedSnapshotShard(Snapshot snapshot, ShardId shardId, String failure, DiscoveryNode masterNode) {
+        sendSnapshotShardUpdate(
+            snapshot, shardId, new ShardSnapshotStatus(clusterService.localNode().getId(), State.FAILED, failure), masterNode);
     }
 
     /** Updates the shard snapshot status by sending a {@link UpdateIndexShardSnapshotStatusRequest} to the master node */
-    void sendSnapshotShardUpdate(final Snapshot snapshot,
-                                 final ShardId shardId,
-                                 final ShardSnapshotStatus status,
-                                 final DiscoveryNode masterNode) {
+    void sendSnapshotShardUpdate(Snapshot snapshot, ShardId shardId, ShardSnapshotStatus status, DiscoveryNode masterNode) {
         try {
             if (masterNode.getVersion().onOrAfter(Version.V_6_1_0)) {
```

Comment:
I simply made it a hard condition here between the old and the new path, the old path doesn't use the de-duplicator. 
I figured the optimization isn't really important in the rolling upgrade case and this makes future backports a lot easier than having the conditionals for that in the old code, right?

Reply:
++